mostlyai/sdk/_data/non_context.py

```diff
@@ -621,27 +513,122 @@ def pull_fk_model_training_data(
         parent_table: Parent table
         tgt_parent_key: Foreign key column in target table
         schema: Schema to fetch context parent data for tgt_table
+        max_children_per_parent: Maximum children to keep per parent
+        max_children: Maximum total children to keep for training
 
     Returns:
         Tuple of (parent_data, tgt_data). tgt_data includes columns from context parent table.
     """
-    parent_data = fetch_parent_data(parent_table=parent_table)
+    t0 = time.time()
+    parent_primary_key = parent_table.primary_key
+    assert parent_primary_key is not None
 
-    parent_keys = parent_data[parent_table.primary_key].tolist() if not parent_data.empty else []
-    tgt_data = fetch_tgt_data(
-        tgt_table=tgt_table,
-        tgt_parent_key=tgt_parent_key,
-        parent_keys=parent_keys,
-        schema=schema,
+    # Step 1: Pull ALL parent keys from children (we only need the FK column for filtering)
+    all_children_keys = tgt_table.read_data(columns=[tgt_parent_key], do_coerce_dtypes=True)
+
+    # Step 2: Pull ALL parent keys
+    all_parent_keys = (
+        parent_table.read_data(columns=[parent_primary_key], do_coerce_dtypes=True)[parent_primary_key]
+        .dropna()
+        .unique()
+    )
+    all_parent_keys_set = set(all_parent_keys)
+
+    # Step 3: Filter children to max_children_per_parent
+    # Remove rows where parent key is null or not in parent table
+    all_children_keys = all_children_keys[
+        all_children_keys[tgt_parent_key].notna() & all_children_keys[tgt_parent_key].isin(all_parent_keys_set)
+    ]
+    filtered_children_keys = all_children_keys.groupby(tgt_parent_key, as_index=False).head(max_children_per_parent)
+
+    # Step 4: Count remaining children
+    num_children = len(filtered_children_keys)
+
+    # Step 5: Keep max_children for training
+    if num_children > max_children:
+        filtered_children_keys = filtered_children_keys.sample(n=max_children, random_state=42).reset_index(drop=True)
+        num_children = max_children
```

Comment:
We don't use `num_children` anywhere afterwards. Might make sense merging step 4&5 into one